content/4_Gold/Intro_DP.mdx

```diff
@@ -242,30 +229,140 @@ public class Main {
 
 ```py
 N = int(input())
-# height is 1-indexed so it can match up with dp
-height = [0] + [int(s) for s in input().split()]
-assert N == len(height) - 1
-
+height = [int(s) for s in input().split()]
 """
 dp[N] is the minimum cost to get to the Nth stone
 (we initially set all values to INF)
 """
-dp = [float("inf") for _ in range(N + 1)]
-# dp[1] = 0 is our base case since we're already at the first stone
-dp[1] = 0
-
-for i in range(1, N + 1):
-	if i + 1 <= stone_num:
-		dp[i + 1] = min(dp[i + 1], dp[i] + abs(height[i] - height[i + 1]))
-	if i + 2 <= stone_num:
+dp = [float("inf") for _ in range(N)]
+# dp[0] = 0 is our base case since we're already at the first stone
+dp[0] = 0
+
+for i in range(1, N - 1):
+	dp[i + 1] = min(dp[i + 1], dp[i] + abs(height[i] - height[i + 1]))
+	if i + 2 < N:
 		dp[i + 2] = min(dp[i + 2], dp[i] + abs(height[i] - height[i + 2]))
+print(dp[N - 1])
+```
+
+</PySection>
+</LanguageSection>
+
+</Spoiler>
+
+<Spoiler title="Pull DP">
+
+There are two ways to get to stone $i$: from stone $i - 1$ and stone $i - 2$.
+- Jump from stone $i - 1$, incurring a cost of $|\text{height}_i - \text{height}_{i-1}|$:
+
+  $$
+  \texttt{dp}[i] = \min(\texttt{dp}[i], \texttt{dp}[i - 1] + |\text{height}_i - \text{height}_{i - 1}|)
+  $$
+
+- Jump from stone $i - 2$, incurring a cost of
+  $|\text{height}_i - \text{height}_{i - 2}|$:
+  $$
+  \texttt{dp}[i] = \min(\texttt{dp}[i], \texttt{dp}[i - 2] + |\text{height}_i - \text{height}_{i - 2}|)
+  $$
+
+**Base case:** $\texttt{dp}[0] = 0$ (start at stone 0)
```

Comment:
if you're gonna add a base case header here, probably do it w/ the push dp asw

Reply:
asw? 😅

Reply:
*as well, sorry